pyasice/ocsp.py

```diff
@@ -0,0 +1,223 @@
+        ("extn_value", OctetString),  # This replaces ParsableOctetString
+    ]
+
+
+class OCSPError(PyAsiceError):
+    pass
+
+
+class OCSP(object):
+    """
+    Certificate validation request via the OCSP protocol, using the asn1crypto/ocspbuilder stack.
+
+    Perform certificate validation:
+
+        ocsp = OCSP(ocsp_service_url)
+        ocsp.validate(subject_cert, issuer_cert, signature)
+
+    Check an arbitrary response:
+
+        ocsp = OCSP.load(der_encoded_binary_data)
+    """
+
+    REQUEST_CONTENT_TYPE = "application/ocsp-request"
+    RESPONSE_CONTENT_TYPE = "application/ocsp-response"
+
+    def __init__(self, url=None):
+        """"""
+        self.url = url
+        self.ocsp_response: Optional[OCSPResponse] = None
+
+    def validate(self, subject_cert, issuer_cert, signature):
+        """
+        A correct OCSP request for BDOC-TM shall contain a DER-encoded "Hash value of the binary value of
+        the signature along with hash algorithm identifier" in the `nonce` field. Yet at the moment, apparently
+        the OCSP service doesn't qualify for time-mark BDOCs so this is probably not significant
+
+        Provide certificates as asn1crypto.x509.Certificate object, oscrypto.asymmetric.Certificate object,
+          a buffer (bytes) in PEM or DER format, or a path to a certificate file.
+
+        :param issuer_cert: a certificate object, a buffer or a path to file
+        :param subject_cert: a certificate object, a buffer or a path to file
+        :param bytes signature: required for BDOC-TM format
+        :return:
+        """
+        ocsp_request = self.build_ocsp_request(subject_cert, issuer_cert, signature)
+
+        try:
+            response = requests.post(
+                self.url,
+                data=ocsp_request.dump(),
+                headers={"Content-Type": self.REQUEST_CONTENT_TYPE, "Connection": "close"},
+            )
+            response.raise_for_status()
+        except requests.ConnectionError:
+            raise OCSPError(f"Failed to connect to OCSP service at {self.url}")
+        except requests.HTTPError as e:
+            raise OCSPError(f"Bad response from OCSP service at {self.url}: {e}")
+
+        content_type = response.headers["Content-Type"]
+        if content_type != self.RESPONSE_CONTENT_TYPE:
+            raise OCSPError(f"Invalid response content type '{content_type}' returned by OCSP service at {self.url}")
+
+        ocsp_response = OCSPResponse.load(response.content)
+        self.verify_response(ocsp_response)
+
+        self.ocsp_response = ocsp_response
+        return self
+
+    def get_responder_certs(self) -> Tuple[ASN1Certificate]:
+        """Get OCSP responder certificates embedded in the response"""
+        return tuple(self.ocsp_response.basic_ocsp_response["certs"])
+
+    def get_encapsulated_response(self):
+        """Get a DER-encoded OCSP response"""
+        return self.ocsp_response.dump()
+
+    def verify(self):
```

Comment:
The thing is, they are all used externally on different occasions =) I agree it's a bit of confusion but after all, you validate certificates and verify responses... I could've named `validate` `validate_certificate` but it's also kinda ugly as the parameter most likely contains `cert` in it, so wished to avoid `butter is butter`